tests/41end-to-end-keys/08-cross-signing.pl

```diff
@@ -682,11 +682,21 @@
       })->then( sub {
          sync_until_user_in_device_list( $user1, $user2 );
       })->then( sub {
-         matrix_get_e2e_keys( $user1, $user2_id );
+         retry_until_success {
+            # Wait until user1 sees signatures uploaded by user2.
+            # We retry_until_success because this is sensitive to replication delays.
+            matrix_get_e2e_keys( $user1, $user2_id )->then( sub {
+               my ( $content ) = @_;
+               log_if_fail "key query content2", $content;
+               $content->{device_keys}->{$user2_id}->{$user2_device}->{"signatures"}
+                  or die "No 'signatures' key present";
```

Comment:
is it sufficient to wait for *anything* at `$content->{device_keys}{$user2_id}{$user2_device}`, or do we need to wait for there to be a `signatures` element?

Reply:
As I understand it, we need to wait for the signatures in particular. 

On `server1:`
1. user2 joins a room
2. user2 uploads signatures

On `server0`, user1 syncs until they see user2's device. This is racey: the sync may complete before the signatures have uploaded, propagated over federation and then over replication.

Reply:
right. please could you add these words to a comment in the code.